public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/PyroscopeAppIntegration.tsx

```diff
@@ -0,0 +1,133 @@
+import { getBackendSrv } from '@grafana/runtime';
+import {LinkButton} from '@grafana/ui'
+
+import { PyroscopeDataSource } from '../datasource';
+import { PyroscopeDataSourceOptions, Query } from '../types';
+
+const PYROSCOPE_APP_ID = 'grafana-pyroscope-app'
+
+/* Global promise to fetch the pyroscope app settings */
+let pyroscopeAppSettings: Promise<PyroscopeAppSettings> | null= null;
+/* Global promises to fetch pyroscope datasource settings by uid as encountered */
+const pyroscopeDatasourceSettingsByUid: Record<string, Promise<PyroscopeDatasourceSettings>> = {};
+
+export type Props = QueryEditorProps<PyroscopeDataSource, Query, PyroscopeDataSourceOptions>;
+
+/** A subset of the app settings that are relevant for this integration */
+type PyroscopeAppSettings = {
+  enabled: boolean;
+  jsonData: {
+    backendUrl: string;
+    basicAuthUser: string;
+  }
+}
+
+/** A subset of the datasource settings that are relevant for this integration */
+type PyroscopeDatasourceSettings = {
+  url: string;
+  basicAuthUser: string | number; // This might be configured as a number, so we will want to ensure compare for string-equivalence
+}
+
+export function PyroscopeAppIntegration(props: Props) {
+
+  const [appPlugin, setAppPlugin] = useState<PyroscopeAppSettings>()
+  const [datasource, setDatasource] = useState<PyroscopeDatasourceSettings>()
+
+  const {datasource:{uid: datasourceUid}, query, range} = props;
+
+  useEffect(()=>{
+
+    if (pyroscopeAppSettings == null) {
+      pyroscopeAppSettings = getBackendSrv().get<PyroscopeAppSettings>(`/api/plugins/${PYROSCOPE_APP_ID}/settings`)
+    }
+
+    pyroscopeAppSettings.then(setAppPlugin);
+    
+  }, [])
+
+  useEffect(()=>{
+    let datasourceSettings = pyroscopeDatasourceSettingsByUid[datasourceUid];
+
+    if (datasourceSettings == null) {
+      datasourceSettings = getBackendSrv().get<PyroscopeDatasourceSettings>(`/api/datasources/uid/${datasourceUid}`)
+      pyroscopeDatasourceSettingsByUid[datasourceUid] = datasourceSettings;
+    }
+
+    datasourceSettings.then(setDatasource);
+
+  }, [datasourceUid])
+
+  const queryParam = useMemo(()=>{
+
+    return generateQueryParams(query, range);
+  }, [query, range])
+
+  const profilesAppLink = useMemo(()=>isPyroscopeDatasourceCompatibleWithPlugin(datasource, appPlugin) &&
+    <LinkButton variant='secondary' icon='external-link-alt' tooltip={'Open query in Profiles App'} target='_blank' href={`/a/${PYROSCOPE_APP_ID}/single?${queryParam}`}>Profiles App</LinkButton>
+  , [appPlugin, datasource, queryParam])
+
+  return profilesAppLink || null;
+}
+
+export function isPyroscopeDatasourceCompatibleWithPlugin(datasource?: PyroscopeDatasourceSettings, appPlugin?: PyroscopeAppSettings) {
+  if (!appPlugin || !datasource) {
+    return false;
+  }
+
+  return (
+    datasource.url === appPlugin.jsonData.backendUrl
+    &&
+    String(datasource.basicAuthUser) === String(appPlugin.jsonData.basicAuthUser)
```

Comment:
How is the app configured? Why doesn't it just reference an existing data source? Could we change that cause repeating the config does not make much sense to me.

Reply:
Yes, the configuration for the app plugin is very close to the content of what is in the datasource.
I did some experimenting earlier, and you are right -- we could have the app use the datasource proxies directly.
This means:
- The pyroscope app would be able to select between multiple pyroscope datasources
- If we accomodated the previous point, we wouldn't need to do this compatibility check -- we would just need to make the selected datasource be part of the URL parameters

Reply:
> Would that be a lot of work right now?

It isn't near on our road map. If you think removing it will increase the likelihood of this getting accepted and merged, I can work without it and use an assumption to determine whether or not to show the button. This assumption would be made on the plugin extension side, keeping the extension point clean.